[로또] 김수환 과제 제출합니다

genius00hwan opened the pull request:
아직 예외처리에 대한 코드가 부족합니다.
모든 함수를 userController 클래스에 다 때려박았습니다. 꾸준히 리팩토링 하도록 하겠습니다.

genius00hwan committed:
test

genius00hwan committed:
docs(README.md): fix README file

genius00hwan committed:
docs(README.md): fix README file

genius00hwan committed:
feat(UserController): creat UserController class

genius00hwan committed:
feat(UserController): create Constructor and inputWinningNumber Method

genius00hwan committed:
feat(UserController): add 'purchase a lotto' method

genius00hwan committed:
add(UserController): add purchaseLotto method and compare method

genius00hwan committed:
fix(UserController): fix compare method

genius00hwan committed:
add(UserController): add rank method

genius00hwan committed:
add(UserController): add initReword Method

genius00hwan committed:
fix(UserController): fix rank method

genius00hwan committed:
add(UserController): add checkUserLotto method

genius00hwan committed:
add(UserController): add isInteger method for validate input String

genius00hwan committed:
fix(UserController): add Exception handling logic

genius00hwan committed:
fix(UserController): fix Readme

genius00hwan committed:
fix(UserController): fix inputWinningNumbers method

KeonHee commented on src/test/java/lotto/ApplicationTest.java:
```diff
@@ -10,52 +10,52 @@
-    }
+	@Test
+	void 기능_테스트() {
+		assertRandomUniqueNumbersInRangeTest(
```
여기는 뭔가 추가된거 없이 indent만 늘어난 거 같은데요?

KeonHee commented on src/main/java/lotto/Application.java:
```diff
@@ -3,5 +3,6 @@
 public class Application {
     public static void main(String[] args) {
         // TODO: 프로그램 구현
+        new UserController();
```
객체 지향 관점에서 클래스는 각각 아래와 같은 역할을 수행한다고 보면되는데요.(단순하게)
- 클래스 생성 -> 변수와 함수를 생성
- 변수 -> 상태 저장
- 함수 -> 무엇을 실행


여기서는 클래스 생성만하면 게임이 실행되죠? 초기화 및 함수 실행이 동시에 진행되는건데요.
생성자에는 변수 초기화 외에는 다른 로직이 안들어가는 것이 좋습니다.
아래처럼 변수 초기화와 실행을 나눠보세요.
```java
UserController controller = new UserController();
controller.run();
```

KeonHee commented on src/main/java/lotto/UserController.java:
```diff
@@ -0,0 +1,155 @@
+		winningLotto = new Lotto(winningNums);
+		inputBonusNum();
+		checkUserLotto();
+		printResult();
```
이런 함수들은 클래스 밖에서 주입받아서 사용할 수 있을 거 같은데요. 
저번 과제나 다른 사람들 코드 참고해서 한번 추상화 시켜보세요~

genius00hwan replied:
네 피드백 감사합니다! 꾸준히 리팩토링 하도록 하겠습니다.

KeonHee reviewed (changes requested):
추상화 관점에서 리뷰했습니다.

추상화가 잘되면 자동으로 테스트 짜기 편한 코드가 됩니다.
수정 후에 유닛 테스트도 추가해보세요~

genius00hwan committed:
fix(LottoMachine): create LottoMachine class

genius00hwan committed:
feat(Machine): create Machine Interface
feat(SettingValues): create SettingValues enum class

genius00hwan committed:
feat(Machine): create AutoLottoMachine Class

genius00hwan committed:
add(SettingValues): add values and constructor for SettingValues

genius00hwan committed:
fix( UserController ): refactor

genius00hwan committed:
fix(Application): construct constructor of UserController

genius00hwan committed:
feat(LottoCompareMachine): create LottoCompareMachine

genius00hwan committed:
feat(Messages): create Messages Enum Class for Print messages

genius00hwan committed:
feat(Messages): create Printer Class for Print messages

genius00hwan committed:
fix(UserController): fix UserController to refactor

genius00hwan committed:
feat(Validator): create Validator for validate input

genius00hwan committed:
add(printer): add printResult() for printResult

genius00hwan committed:
feat(RewardMeasuringMachine): create RewardMeasuringMachine for measurement user's reward

genius00hwan committed:
add(RewardMeasuringMachine): add constructor and rank method

genius00hwan committed:
add(SettingValues): add First - Fifth values

genius00hwan committed:
fix(UserController): refactoring

genius00hwan committed:
feat(SharingVariables): create Sharing Values

genius00hwan committed:
add(SettingValues) : add point values

genius00hwan committed:
fix(Messages) : fix winning message format

genius00hwan committed:
add(printer) : add methods for input messages and divide the printResult func

genius00hwan committed:
fix(LottoCompareMachine) : using printer

genius00hwan committed:
fix(RewardMeasuringMachine) : fix the logics

genius00hwan committed:
fix(UserController) : refactoring

genius00hwan committed:
add(SharingVariables) : add values

genius00hwan committed:
add(ErrorMessages) : create ErrorMessages EnumClass

genius00hwan committed:
add(Validator) : add isInvalidUnit func and isThereSapce func

genius00hwan committed:
feat(JudgeMachine) : create JudgeMachine interface

genius00hwan committed:
feat(NumberCompareMachine) : create NumberCompareMachine interface

genius00hwan committed:
refactor(Packages) : refactoring

genius00hwan committed:
refactor(Packages) : refactoring

genius00hwan committed:
refactor(Packages) : refactoring

genius00hwan committed:
feat(Validator) : creat validator interface

genius00hwan committed:
refactor(Packages) : refactoring

genius00hwan committed:
feat(ErrorMessages) : creat ErrorMessages Enum class

genius00hwan committed:
refactoring

genius00hwan committed:
rewrite Readme

genius00hwan reviewed (commented):
리팩토링 했습니다. 아직 테스트 코드는 추가하지 못했습니다.

genius00hwan committed:
feat(RandomNumsGenerator) : create RandomNumsGenerator interface

genius00hwan committed:
fix(AutoLottoMachine) : divide logic for implement RandomNumsgenerator

genius00hwan committed:
add(Printer) : add printList func for print List<>

genius00hwan committed:
change java version to 17

genius00hwan committed:
feat(UserValues) : move Values for user from Lotto config

KeonHee commented:
@genius00hwan 어떻게 리팩토링 했는지 설명해줄 수 있나요?

genius00hwan reviewed (commented):
config 패키지에 설정값, 메시지 등의 상수를 넣었습니다.
lotto 패키지에는 user'slotto등의 user 전용 값, 로또 구매를 하고 다른 함수를 실행 시킬 수 있는 controller을 넣었습니다. 또, printer 는 프로그램 사용시에 메시지를 출력합니다.
machine 패키지에는 자동로또를 생성하는 machine, user lotto를 winning lotto와 비교하는 machine, 비교결과로 등수를 매기는 machine이 들어가 있습니다. 각 machine들은 랜덤숫자생성기의 interface, 채점을 하는 interface, 등수를 매기는 interface를 구현한 결과 입니다.
validate 패키지에는 error message와 validator interface가 들어가 있고 LottoValidator는 lotto게임 전용으로 validator를 구현한 것입니다.

genius00hwan committed:
fix(UserController) : move Values for user from UserValues

genius00hwan reviewed (commented):
cmd + option + L 로 가독성을 높였습니다.
userValue라는 클래스로 User의 고유값을 사용했다가 UserController 클래스로 옮겼습니다.

genius00hwan committed:
fix(...) : add try-catch for handle exception
fix(RewardMeasuringMachine) : userLottos initialize in constructor and substitute at run()

genius00hwan committed:
add(ApplicationTest.java) : add test code

genius00hwan committed:
fix(LottoCompareMachine) : remove debug code

genius00hwan committed:
add(Validator) : add PrintErrorMessage function

genius00hwan committed:
add(Application) : add test codes

genius00hwan committed:
feat(NumConverter) : create NumConverter

genius00hwan committed:
fix(Printer) : fix using NumConverter

genius00hwan committed:
reFact(NumberGame) : create numberGame package and move interfaces to the package

genius00hwan committed:
add(LottoConfig) : add Cut_size value

genius00hwan committed:
fix(Messages) : remove string value of rewards and using NumConverter

genius00hwan committed:
fix() : packaging has changed

genius00hwan reviewed (commented):
숫자를 문자열로 변환하는 로직을 추가했습니다.
L 1000단위로 콤마를찍게 했습니다.
   소수부는 주어진 자리수 까지 반올림 하게했습니다.
numberGame 패키지를 추가해 이 패키지로 다른 숫자게임을 만들 수 있게 재사용성을 고려했습니다.

genius00hwan committed:
fix(userController) : fix for dependency injection
